tests/test_manager.py

```diff
@@ -372,3 +372,98 @@ def test_request_body_format(self):
         body = manager.save_or_put({"bing": "bong"})[3]
 
         self.assertTrue(body, "<Invoice><bing>bong</bing></Invoice>")
+
+    def test_idempotency_key(self):
+        """
+        A valid idempotency key should be provided as a header on post and
+        put API calls. An invalid idempotency key should raise an exception.
+        """
+        from io import BytesIO
+
+        from xero.utils import generate_idempotency_key
+
+        credentials = Mock(base_url="", user_agent=None)
+        manager = Manager("Invoices", credentials)
+
+        with patch("xero.basemanager.requests.post") as mock_post:
+            # Don't provide an idempotency key so we should not
+            # have a header in the request
+            with self.assertRaises(KeyError):
```

Comment:
This method reads as a series of at least 6 different tests, not 1. Tests should be as granular as possible, so that the set of test failures gives a better indication of the source of any problem. If there's common (and complex) setup/teardown logic, that can be factored out and re-used between tests.